stdlib/public/core/UnicodeHelpers.swift

```diff
@@ -0,0 +1,377 @@
+internal func _utf8ScalarLength(
+  _ utf8: UnsafeBufferPointer<UInt8>, endingAt i: Int
+  ) -> Int {
+  var len = 1
+  while _isContinuation(utf8[i &- len]) {
+    len += 1
+  }
+  _sanityCheck(len == _utf8ScalarLength(utf8[i &- len]))
+  return len
+}
+
+@inlinable @inline(__always)
+internal func _isContinuation(_ x: UInt8) -> Bool {
+  return x & 0b1100_0000 == 0b1000_0000
+}
+
+@inlinable
+@inline(__always)
+internal func _continuationPayload(_ x: UInt8) -> UInt32 {
+  return UInt32(x & 0x3F)
+}
+
+@inline(__always)
+internal func _decodeSurrogatePair(
+  leading high: UInt16, trailing low: UInt16
+) -> UInt32 {
+  _sanityCheck(_isLeadingSurrogate(high) && _isTrailingSurrogate(low))
+  let hi10: UInt32 = UInt32(high) &- UInt32(_leadingSurrogateBias)
+  _sanityCheck(hi10 < 1<<10, "I said high 10. Not high, like, 20 or something")
+  let lo10: UInt32 = UInt32(low) &- UInt32(_trailingSurrogateBias)
+  _sanityCheck(lo10 < 1<<10, "I said low 10. Not low, like, 20 or something")
+
+  return ((hi10 &<< 10) | lo10) &+ 0x1_00_00
+}
+
+@inline(__always)
+internal func _numUTF8CodeUnits(_ scalar: Unicode.Scalar) -> Int {
+  switch scalar.value {
+    case 0..<0x80: return 1
+    case 0x80..<0x0800: return 2
+    case 0x0800..<0x1_0000: return 3
+    default: return 4
+  }
+}
+@inline(__always)
+internal func _numUTF16CodeUnits(_ scalar: Unicode.Scalar) -> Int {
+  return scalar.value <= UInt16.max ? 1 : 2
+}
+
+//
+// Scalar helpers
+//
+extension _StringGuts {
+  @usableFromInline @inline(__always) // fast-path: fold common fastUTF8 check
+  internal func scalarAlign(_ idx: Index) -> Index {
+    // TODO(String performance): isASCII check
+
+    if _slowPath(idx.transcodedOffset != 0 || idx.encodedOffset == 0) {
+      // Transcoded indices are already scalar aligned
+      return String.Index(encodedOffset: idx.encodedOffset)
+    }
+    if _slowPath(self.isForeign) {
+      return foreignScalarAlign(idx)
+    }
+
+    return self.withFastUTF8 { utf8 in
+      var i = idx.encodedOffset
+      while _slowPath(_isContinuation(utf8[i])) {
+        i -= 1
+        _sanityCheck(
+          i >= 0, "Malformed contents: starts with continuation byte")
+      }
+      // If no alignment is performed, keep grapheme cache
+      if i == idx.encodedOffset {
+        return idx
+      }
+
+      return Index(encodedOffset: i)
+    }
+  }
+
+  @inlinable
+  internal func fastUTF8ScalarLength(startingAt i: Int) -> Int {
+    _sanityCheck(isFastUTF8)
+    let len = _utf8ScalarLength(self.withFastUTF8 { $0[i] })
+    _sanityCheck((1...4) ~= len)
+    return len
+  }
+
+  @inlinable
+  internal func fastUTF8ScalarLength(endingAt i: Int) -> Int {
+    _sanityCheck(isFastUTF8)
+
+    return self.withFastUTF8 { utf8 in
+      _sanityCheck(i == utf8.count || !_isContinuation(utf8[i]))
+      var len = 1
+      while _isContinuation(utf8[i - len]) {
+        _sanityCheck(i - len > 0)
+        len += 1
+      }
+      _sanityCheck(len <= 4)
+      return len
```

Comment:
I believe it is rather important to explicitly check for `len <= 4` in all cases, rather than this being a compile-dependent check. Some dependent code using `fastUTF8ScalarLength` does `Builtin.unreachable()` if length is not one of `{1,2,3,4}`. Therefore the program might crash on some invalid input if the length is not properly bound to `[1..4]` range over here.

Reply:
This is only for native strings, which we validate on creation